### PR TITLE
#1401. [Patterns] Pattern assignment tests

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Patterns/pattern_assignment_A05_t06.dart [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    af6 = 0;
+    af7 = 0;
+    af8 = 0;
+    areaAsInt = 0;
+    rf = ();
+    lf = [0];
+  }
+
+  Expect.throws(() {
+    ((af1, _) && r) = (1, 2);
```
Comment (chloestefantsova) [inline LanguageFeatures/Patterns/pattern_assignment_A05_t06.dart:53]: It seems that this line should trigger a compile-time error. I wrote the following program to see the behavior of both the CFE and the analyzer in analogous situation without patterns, and both tools report an error.

```dart
test(void Function() f) => f();

main() {
  final int a;
  if (2 > 1) {
    a = 0;
  }
  test(() {
      a = 1;
  });
}
```

The output from the CFE.
```
$ xcodebuild/ReleaseX64/dart /tmp/foo.dart
/tmp/foo.dart:9:7: Error: Final variable 'a' might already be assigned at this point.
      a = 1;
      ^
```

The output from the Analyzer.
```
$ xcodebuild/ReleaseX64/dart analyze /tmp/foo.dart
Analyzing foo.dart...

  error - foo.dart:9:7 - The final variable 'a' can only be set once. Try making 'a' non-final. - assignment_to_final_local
   info - foo.dart:4:13 - The value of the local variable 'a' isn't used. Try removing the variable or using it. - unused_local_variable

2 issues found.
```

Similarly below.

**Review by chloestefantsova — CHANGES_REQUESTED** [review]

File: LanguageFeatures/Patterns/pattern_assignment_A03_t01.dart [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+main() {
+  int a1 = 42;
+  Record r1 = ();
+  ((a1, _) && r1) ??= (1, 2);
+//^^^^^^^^^^^^^^^
+// [analyzer] unspecified
+// [cfe] unspecified
+
+  int a2 = 3;
+  [a2, _] = [1, 2];
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Patterns/pattern_assignment_A03_t01.dart]: Forgot compound operator. Fixed

File: LanguageFeatures/Patterns/pattern_assignment_A02_t01.dart [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  Expect.listEquals(x2, [1, 2]);
+
+  var a3 = 42;
+  var x3 = {"key1": a3, "key2": _} = {"key1": 1, "key2": 2};
+  Expect.equals(1, a3);
+  Expect.mapEquals({"key1": 1, "key2": 2}, x3);
+
+  var a4 = 3;
+  final x4 = (a4, _) = (1, 2);
+  Expect.equals(1, a1);
```
[the remark this answers is not in the document]
Comment (sgrekhov) [reply LanguageFeatures/Patterns/pattern_assignment_A02_t01.dart]: Of course! Fixed, thanx

File: LanguageFeatures/Patterns/pattern_assignment_A05_t06.dart (hunk 1 above)
Comment (sgrekhov) [reply LanguageFeatures/Patterns/pattern_assignment_A05_t06.dart:53]: Test description clearly says about `late final` but the test had only `final`. Added `late`. With the `late` an example above produces a runtime error, not a compile-time. Thank you for noticing

**Review by sgrekhov — COMMENTED** [review]
@chloestefantsova please review one more time

**Review by chloestefantsova — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**